### ISSUE TITLE
`<iframe>` elements stop page from loading

### ISSUE THREAD

**Issue by escritorio-gustavo** [mattsse/chromiumoxide#163]
I am trying to visit a page that uses Google's RecaptchaV3. The issue I'm having is the following: for some reason, the equest sent by the `src` attribute of the `<iframe>` element gets stuck on pending, on the content download phase, causing `page.goto` to fail.

If, for any reason, the rust code panics and the browser survives as a zombie process, everything imediately goes back to normal, i.e. the request resolves as soon as Rust crashes.

Below are screenshots of the problem and a minimal reproduction of the code I'm using:
```rust
use std::{path::Path, time::Duration};

use chromiumoxide::{Browser, BrowserConfig, BrowserFetcher, BrowserFetcherOptions};
use futures::StreamExt;

#[tokio::main]
async fn main() {
    let download_path = Path::new("./download");
    let _ = std::fs::create_dir_all(download_path);

    let fetcher_config = BrowserFetcherOptions::builder()
        .with_path(download_path)
        .build()
        .unwrap();

    let fetcher = BrowserFetcher::new(fetcher_config);

    let info = fetcher.fetch().await.unwrap();

    let config = BrowserConfig::builder()
        .chrome_executable(info.executable_path)
        .launch_timeout(std::time::Duration::from_secs(20))
        .with_head()
        .arg("--lang=en_US")
        .no_sandbox()
        .build()
        .unwrap();

    let (mut browser, mut handler) = Browser::launch(config).await.unwrap();

    let handle = tokio::spawn(async move {
        while let Some(h) = handler.next().await {
            if h.is_err() {
                break;
            }
        }
    });

    browser
        .new_page("https://esaj.tjms.jus.br/cpopg5/open.do")
        .await
        .unwrap();

    tokio::time::sleep(Duration::from_secs(50)).await;

    browser.close().await.unwrap();
    browser.wait().await.unwrap();
    handle.await.unwrap();
}
```

The weird thing about it is that it does have a 200 HTTP status code, even though it's still pending
![image](https://github.com/mattsse/chromiumoxide/assets/131818645/c41daaea-c3b9-4af2-871f-20b84b1602af)

**Comment by escritorio-gustavo** [comment]
Notes:
* I have tried intercepting this request to abort it and it didn't work. The request still gets sent and still gets stuck.
* I have tried intercepting this request to manually fulfill it, providing the HTML as a string to `FulfillRequestParams`. This didn't work either. The request still gets sent and still gets stuck.
* In the reproduction above I didn't use `page.goto`, I opened a new page directly on the URL, this does not change the outcome either.

**Comment by escritorio-gustavo** [comment]
This issue seems to be specifically caused by `iframe` elements, and not necessarally reCaptcha. I have tested it on the following [codepen](https://codepen.io/IanLintner/pen/DqGKQZ), and otained the same result. Loading of the page stops, the maps don't show up, and crashing the rust process gets everything back to normal

**Comment by escritorio-gustavo** [comment]
Please, if anyone can help, I really need this to work

**Comment by chirok11** [comment]
@escritorio-gustavo 
If I understand your issue correctly, the problem is that there's a small bug that isn't being handled in chromiumoxide, a bug in Chromium itself. 
You could try forking the repository yourself, and in the file **src/handler/target.rs**, on line _521_, change "auto_attach" to false and let me know if that solves the problem.

**Comment by escritorio-gustavo** [comment]
@chirok11  This does indeed work, the pages open successfully even in non-headless mode, thank you so much.

Do you have any idea what this chromium bug is and why this auto attach causes it to happen?

Also, does disabling auto attach cause any other side effects (i.e. can I submit a pull request to fix this close this issue without breaking the whole library?)

**Comment by chirok11** [comment]
That could be fixed since other libraries like puppeteer or playwright is working correctly within these flags.
If you wish to fix it, then take a look for target auto attach handling in other libraries. 
I saw links to crbug.com yesterday, but it is not yet solved for a while.
I am trying to discover issue, and supposing now that chrome is not auto attaching to iframes or service workers and that's why they're stuck and disconnect is solved. 
And client should attach manually or disconnect from these targets.

**Comment by escritorio-gustavo** [comment]
@chirok11 

Thank you so much, I've been dealing with this issue for a couple of months now (I've been switching between this crate and headless_chrome, which is FAR slower)

I think I'll submit a pull request that adds this fix as a feature using the `cfg` macro. Again, thanks a lot

**Comment by chirok11** [comment]
@escritorio-gustavo 
Hello, couldn't you please check if this issue with iframe or service workers persist on my branch?
Here is my pull request https://github.com/mattsse/chromiumoxide/pull/173

Take a look here.
https://github.com/puppeteer/puppeteer/blob/main/packages/puppeteer-core/src/common/ChromeTargetManager.ts#L343

Here is workaround for service workers, we should detach for them, also, we have to attach to iframe's manually, but currently I am just telling runIfWaitingForDebugger (otherwise, it will load forever.) because I didn't see function's for on_target_attach in handler Target context.

**Comment by escritorio-gustavo** [comment]
Hey @chirok11, did you close #173 PR because of #170? If so, let me know and I'll close #170 instead, as it is just a workaround using the method you mentioned with setting `auto_attach` to false.

I have absolutely zero knowledge on the subject and I'm willing to bet your PR provides a way better solution

**Comment by chirok11** [comment]
I am still trying to fix it in correct way.
I don't have much knowledge too, but trying to do best.
I forked repository and has a branch cdp-fx. Could you try to reproduce issue on my branch?
I kept auto attaching, but detaching from service workers and added runIfWaitingForDebugger.
Most broken sites loading now, but I found an issue with some sites when goto future will always return timeout even website is loaded successfully.
Anyway looks like iframes should be handled manually. 
I am not quite sure that it will not broke work with iframes (but really dont know does this library could work with iframes currently.)

**Comment by escritorio-gustavo** [comment]
> I forked repository and has a branch cdp-fx. Could you try to reproduce issue on my branch?

Sure thing, I'll start testing right now.

In the meantime I'll leave #170 open. Once you figure out how to fix the problem you can simply delete the feature flag I added,

**Comment by chirok11** [comment]
https://github.com/chirok11/chromiumoxide/tree/cdp-fx
This is tree. Notice that I also changed Browser::connect and added HandlerConfig argument (to provide custom request timeout)

**Comment by escritorio-gustavo** [comment]
I've tested a couple of pages that used to give me trouble and things worked fine. Haven't bumped into the `goto` problem yet

**Comment by escritorio-gustavo** [comment]
Also it seems like not all iframes will trigger the problem, for instance, in this version of the crate [this W3Schools page](https://www.w3schools.com/html/html_iframe.asp) loads fine, but [this CodePen](https://codepen.io/IanLintner/pen/DqGKQZ) doesn't. 

Both pages load with your fork though

**Comment by chirok11** [comment]
Load, and you future is finished? I mean `goto` doesn't fall on timeout?

**Comment by escritorio-gustavo** [comment]
Found the `goto` problem on the [MDN docs for iframe](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe)

**Comment by chirok11** [comment]
Thanks! Will try to work on it.
[https://www.financialexpress.com/](financialexpress.com) also won't load and causes a timeout.

**Comment by chirok11** [comment]
I think there is something that links them, but it's definitely not Service Worker, as it is indeed not on MDN.
There is something common between these two websites, but so far I only see an iframe as a similarity.
I have a feeling that with certain website behavior (the presence of an iframe), the logic in CommandFuture breaks, and it simply does not receive a response that the page has loaded (even though it should, as the page is actually loaded correctly).

**Comment by chirok11** [comment]
OK, problem is in frame lifecycle, some websites does not emit "load" event (bug in chromiumoxide event handling? dunno now).
If we change expected_lifecycle in [src/handler/frame.rs#L617](https://github.com/mattsse/chromiumoxide/blob/main/src/handler/frame.rs#L617) to "networkIdle" instead of "load" we'll have successfully loading pages without failing them on timeout.

**Comment by chirok11** [comment]
Alright, I've figured it out. It turns out that the issue was, first of all, that check_lifecycle was waiting for a load event from the main frame and all child frames. However, on problematic sites, I encountered child frames with url: None, which will never emit a "load" event.
Interestingly, if you increase the wait time to several seconds after creating the page and then navigate to the page, there's a chance that everything will go smoothly; so, the problem is intermittent.
But if you remove the pause between creating the page and navigating, almost a hundred percent of the time there will be a timeout error.
I've slightly modified the check_lifecycle function in src/handler/frame.rs; I've added the following conditions:
lifecycle_events contains "load", or if frame.url is None, at least check for the presence of DOMContentLoaded.

I think there should be an option to choose whether to wait for "load," "networkIdle," or "networkAlmostIdle," similar to other libraries working with the Chrome DevTools Protocol.

**Comment by chirok11** [comment]
@escritorio-gustavo 
You could try to check MDN links or other and confirm that it is not failing by Timeout for now and other pages doesn't get broken.

**Comment by escritorio-gustavo** [comment]
@chirok11

This worked, the newest version of your fork properly loads the MDN page, financialexpress and the pages where I originally encountered the problem, I will close #170 in favor of this implementation

**Comment by escritorio-gustavo** [comment]
Hey @chirok11, can you add "this will fix #163" to your PR's description to link it to this issue? This way when it's accepted the issue should be automatically closed

**Comment by chirok11** [comment]
@escritorio-gustavo done.

**Comment by escritorio-gustavo** [comment]
Awesome!

Also, it seems like #174 contains the edits you made for #172, so #172 can be closed too to reduce the number of PRs, just add "This will fixs #171" as well

Sorry for not seeing this sooner, I was double checking if the edits were actually the same.

Btw, you did an awesome job! Thanks again for all the help with this issue

**Comment by escritorio-gustavo** [comment]
Btw, the issue #171 wasn't linked to the PR, as it requires the keyword in all issues to create the links, so you need to write "Closes #163 and closes #171"

**Comment by chirok11** [comment]
Thank you, modified comment.
